lib/ndr_import/xml/table.rb

```diff
@@ -24,28 +31,114 @@ def footer_lines
+        records_from_xml_line = []
+        augmented_masked_mappings.each do |klass, klass_mappings|
           fields = mapped_line(xml_line, klass_mappings)
+
           next if fields[:skip].to_s == 'true'.freeze
-          yield(klass, fields, index)
+
+          if yield_xml_record
+            records_from_xml_line << [klass, fields, index]
+          else
+            yield(klass, fields, index)
+          end
         end
+        yield(records_from_xml_line.compact) if yield_xml_record
       end
 
       private
 
+      def augment_and_validate_column_mappings_for(line)
+        augment_column_mappings_for(line)
+        validate_column_mappings(line)
+
+        NdrImport::Xml::MaskedMappings.new(@klass, @augmented_columns.deep_dup).call
+      end
+
+      # Add missing column mappings (and column_xpaths) where
+      # repeating sections / data items appear
+      def augment_column_mappings_for(line)
+        # Start with a fresh set of @augmented_columns for each line, adding new mappings as
+        # required for each `line`
+        @augmented_columns       = @columns.deep_dup
+        @augmented_column_xpaths = column_xpaths.deep_dup
+
+        unmapped_nodes(line).each do |unmapped_node|
+          existing_column = find_existing_column_for(unmapped_node.dup)
+          next unless existing_column
+
+          unmapped_node_parts   = unmapped_node_parts(unmapped_node)
+          klass_increment_match = unmapped_node.match(/\[(\d+)\]/)
+          raise "could not identify klass for #{unmapped_node}" unless klass_increment_match
+
+          new_column = NdrImport::Xml::ColumnMapping.new(existing_column, unmapped_node_parts,
+                                                         klass_increment_match[1], line,
+                                                         @klass).call
+          @augmented_columns << new_column
+          @augmented_column_xpaths << build_xpath_from(new_column)
+        end
+      end
+
+      def xml_line_from(line)
+        @augmented_column_xpaths.map do |column_xpath|
+          # Augmenting the column mappings should account for repeating sections/items
+          # TODO: Is this needed now that we removed "duplicated" klass mappings?
+          line.xpath(column_xpath).count > 1 ? '' : line.xpath(column_xpath).inner_text
+        end
+      end
+
+      def find_existing_column_for(unmapped_node)
+        # Remove any e.g. [2] which will be present on repeating sections
+        unmapped_node.gsub!(/\[\d+\]/, '')
+        unmapped_node_parts = unmapped_node_parts(unmapped_node)
+        columns.detect do |column|
+          column['column'] == unmapped_node_parts[:column_name] &&
+            column.dig('xml_cell', 'relative_path') == unmapped_node_parts[:column_relative_path] &&
+            column.dig('xml_cell', 'attribute') == unmapped_node_parts[:column_attribute]
+        end
+      end
+
+      def unmapped_node_parts(unmapped_node)
```

Comment:
This method name is very confusing!
- Many parameters are named `unmapped_node_parts`, and are generally arrays (of length 3, or occasionally 2 in all the test cases).
- Method `unmapped_node_parts` has the same name, but returns a Hash.